add Secure Releases and CVE Management guide (#70)
openjs-foundation/security-collab-space: secure-releases.md

# Secure Releases and CVE Management for the JS Ecosystem

## Secure Releases

> In order for your JS software project to be safe, secure, and reliable, the versions it releases need to follow security best practices. Below are the OpenJS Foundations recommended best practices for secure releases.

Reviewer: (nit)

```suggestion
In order for your JS software project to be safe, secure, and reliable, its releases need to follow security best practices. Below are the OpenJS Foundations recommended best practices for secure releases.
```

> ### Two-factor Authentication

Reviewer: consider expanding to MFA.  also if covering authentication, you may want to consider guidance on things like using a (good) password manager, best practices around storing recovery codes, etc.

> Two-factor authentication (2FA) means having something more than just your password to confirm you're you. The second factor can be a one-time code using an app like Google Authenticator, a hardware key (such as Yubikey or Titan), or a passkey using touch ID, face ID, or something else that's stored locally on your device. Note: do not *ever* use SMS for 2FA unless it's the only mechanism offered, since SMS is vulnerable to SIM swap and number forgery attacks.

Reviewer: "something else" is vague. consider eliminating.. ambiguity can be problematic when discussing security.

> Two-factor authentication (2FA) means having something more than just your password to confirm you're you. The second factor can be a one-time code using an app like Google Authenticator, a hardware key (such as Yubikey or Titan), or a passkey using touch ID, face ID, or something else that's stored locally on your device. Note: do not *ever* use SMS for 2FA unless it's the only mechanism offered, since SMS is vulnerable to SIM swap and number forgery attacks.

Reviewer: "using an app" -> "using a 2FA/MFA authentication app"

also suggest if you are going to mention a particular app/service at all, it's good to mention more than one.  maybe Authy, etc

> Although this extra authentication step can be a mild inconvenience, the benefit for security is overwhelmingly worth it. Both Github and npm will soon require 2FA for every user, so you might as well enable it now.

Reviewer: there's very little left in the year. I'd suggest just changing it now to say the GH already requires it

Author: They don't quite require it yet, though; I'm happy to update the doc after that's changed.

### Preparing for Publish

It can be easy to forget to run steps before publishing. If you have any build steps (transpiling with Babel or TypeScript, emitting types with TypeScript, etc) and/or validation steps (running tests, etc), make sure they're set up in the `scripts` section of `package.json` such that `npm run prepublish` or `npm run prepublishOnly` would invoke them - that way, running `npm publish` in the future will automatically run these steps.

> Publishing to npm can be done in one step (`npm publish`), or two (`npm pack` to generate a tarball, and then later, `npm publish path/to/tarball` to publish). Even if you publish separately with one step, it's a good idea to run `npm pack` in CI to ensure that the steps succeed, and that the tarball is somewhat reproducible. (TODO: do this in my packages' CI and link to it here as an example)

Reviewer: just reminder of the `TODO` here

Author: Fixed, thanks.

### Publishing

> If you have only 1 person who can publish releases, then this step is optional. If you have 2 or more, this step is strongly recommended.

Reviewer: ```suggestion
If you have only one person whom's't've can publish releases, then this step is optional. If you have two or more, this step is strongly recommended.
```

Author: lol, whom's't've?

> Publish a packed tarball from CI, only upon an explicit request. This means, don't make a new release automatically, but make sure that a human has intentionally said "it's time to publish a new version, here's the version number it will use". This explicit request can be a pushed semver-versioned git tag, or an explicit user-run workflow - but does not include merging/committing to the default branch.

Reviewer: ```suggestion
Publish a packed tarball from CI, only upon an explicit request. This means, don't make a new release automatically, but make sure that a human has intentionally said "it's time to publish a new version, here's the version number it will use". This explicit request can be a pushed SemVer-conformant git tag, or an explicit user-run workflow - but does not include merging/committing to the default branch.
```

Author: I think this correction would actually be incorrect, since a leading `v` isn't conformant with SemVer, but it's the most common idiom in a semver-versioned git tag. Additionally, in a monorepo, git tags are often something like `subpackage@1.2.3`, which is also not Semver-conformant.

Reviewer: I'm confused by this response. All I meant was that the short form of  Semantic Versioning is SemVer, not Semver and not semver.

Reviewer: oh perhaps we are getting wires crossed because I'm suggesting two things:
semver -> SemVer
and
versioned -> conformant

Author: right. re capitalizing semver, i'm not actually referring to the spec, I'm referring to the `semver` npm module, which is the actual source of truth for the JS ecosystem. re "conformant", for the reasons indicated, versioned tags are almost never conformant with the spec.

Reviewer: ok.  I don't have strong feelings here

> Do not publish without using 2FA. Note that an npm automation token is a single factor, and does not qualify as 2FA. Since Github does not provide any functionality to allow you to input a 2FA code natively, you'll need to use something like [Wombat Dressing Room](https://opensource.googleblog.com/2020/01/wombat-dressing-room-npm-publication_10.html) which requires you to host a service on GCP ([example](https://github.com/mathiasbynens/String.prototype.endsWith/blob/main/.github/workflows/publish-on-tag.yml#L21-L22)), or [Step Security's Wait For Secrets](https://github.com/step-security/wait-for-secrets) ([example](https://github.com/jsx-eslint/eslint-plugin-react/blob/28f9a6cdf716bc34f824ea3a3988e68f3e54eb0f/.github/workflows/npm-publish.yml)) which requires you to trust Step Security's external service.

Reviewer 2: Is there anything that the Foundation can do to help maintainers implement the automate 2FA more easily?

Author: Short of maintaining a service for it, or pressuring Github to support it natively, I'm not sure what anyone can do to make it easier.

Reviewer 2: > Short of maintaining a service for it
Is that at all reasonable as a value add by the Foundation for OpenJS projects?

Author: Yes, but the costs might be quite large over time - Foundation budget would be better spent paying Github to implement it natively, i suspect.

## CVE Management

### Security Policy

> Your software project will, likely inevitably, have a security vulnerability. When (not if) that happens, you want to make sure that whoever discovers it has a way to report it to you privately - this is called "responsible disclosure" or "coordinated disclosure". So, the first important step is to *have a security policy*.

Reviewer: (nit)

```suggestion
Your software project will, likely inevitably, have a security vulnerability. When that happens, you want to make sure that whoever discovers it has a way to report it to you privately - this is called "responsible disclosure" or "coordinated disclosure". So, the first important step is to *have a security policy*.
```

Author: Why? the emphasis with the parenthetical is intentional.

Reviewer: it's a nit, but the 'why' answer is because you've already made the point clear

Author: ah, fair enough.

On Github, this is always located at `$REPO_URL/security/policy` ([example](https://github.com/nvm-sh/nvm/security/policy)). This page can be populated from a number of file locations: `SECURITY.md` or `.github/SECURITY.md` in the repo itself, or, one of the same files in your user or organization’s `.github` repository. If you're not on Github, then look for a similar platform-specific convention, and if you find none, a top-level `SECURITY.md` is safe.

In this policy, you'll want to mention a private reporting channel. This could be an email address, a service like HackerOne, etc.

#### Programmatic Private Vulnerability Reporting

> Github and Gitlab support a Private Vulnerability Reporting (PVR) feature and API. This allows reporters to use the platform itself, programmatically if needed. It is strongly recommended to enable this feature, if it's not already on by default.

Reviewer: perhaps mention this can be enabled at the org level for all repos, as well as a setting to automatically enable it for any new repos in the org

### Audit Dependencies

You'll want to ensure all of your dependencies - direct and transitive - are audited on every commit. This can be achieved locally and in CI with `npm audit`, if your project has a lockfile, or https://npmjs.com/aud for packages without a lockfile.

In addition - not instead - you may want to use services like Dependabot, Renovate, [Tidelift](https://tidelift.com), [Socket.dev](https://socket.dev), etc to automatically detect and warn you about issues with your dependency graph.

If your package does not have a build process - meaning, all of the artifacts you publish are committed to version control and hand-written - then you may only need to audit your runtime dependencies (`dependencies`, `peerDependencies`) and not `devDependendies`. However, if you do have a build process, whichever dev deps that generate code need to be audited as well.

### Handling Reports

> First, be highly responsive. Ideally, you can respond to a reporter within 24 hours or 1 business day. Your security policy should explicitly describe your expected response time - especially if it's longer than this.

Reviewer 2: I'm a bit torn on this. Making people feel bad if they don't respond within a day is great for people using the package for free but not so great for maintainers.  I personally think what is important is to state what kind of response time to expect and that it reflect what the maintainer will be realisticly be able to achieve.

If we want a higher bar for this for OpenJS projects, then we might want to include that in one of the entry/application checklists.

Reviewer 2: or I guess offer some support in achieving a fast turn around.

Author: I used "ideally" so that they can set any time they like, as long as it's in the policy.

How would you suggest I rephrase it where I'm still establishing a norm of responsiveness?

Reviewer 2: How about this

```
It is important to set expectations for how quickly you will be able to respond to reporters by documenting this in your security policy. It is more imporant to set a realistic timeframe and meet it than setting a short timeframe that you cannot meet. Once you set the timeframe in your security policy work to meet that consistently.

Reviewer 2: If this guide is mean to be specific to OpenJS projects versus a general guide we could then add a requirement along the lines of something like "As an OpenJS foundation project the requirement is that the response timeframe be 1 week or less" or whatever the requirement is.

Author: It's meant to cover both foundation projects and general JS ecosystem projects.

I don't think there should be any requirement here, necessarily - but I think a business day is a realistic default for security reports.

Reviewer 2: Even within a large project like Node.js we don't make 1 day. This is what we've ended up with:

Normally your report will be acknowledged within 5 days, and you'll receive a more detailed response to your report within 10 days indicating the next steps in handling your submission. These timelines may extend when our triage volunteers are away on holiday, particularly at the end of the year.

Author: I would expect it's *harder* in a large project to have faster response times - but most open source projects are small, and it should be pretty easy to respond within a day.

I absolutely don't want to shame or preclude longer response times, but 5-10 days would be a pretty long time imo for a small project.

Reviewer 2: Assuming somebody has time to check their project every single day when its not their paid job seems a bit much to ask, but at this point I've shared my opinion and I'll bow out of the dicussion to let whatever the consensus is be used.

> First, be highly responsive. Ideally, you can respond to a reporter within 24 hours or 1 business day. Your security policy should explicitly describe your expected response time - especially if it's longer than this.

Reviewer: ```suggestion
First, be highly responsive. Respond to a report as soon as possible; ideally, within 24 hours or one business day. Your security policy should explicitly describe your expected response time - especially if it's longer than this.
```

Author: why 24 but "one"? generally i keep numbers as digits or words consistent throughout the document.

Reviewer: most writing style guides spell out number one through nine or ten. some even continue spelling them out to 100. but generally they agree that single digits get spelled out

Author: not sure if we need to follow any extant writing style guides, but i'll make the change

> When responding to a report, you'll want to seriously consider how the behavior may or may not be a vulnerability. Things to weigh include what documentation says, typical community expectations, what the likely attack surface looks like, etc. This stuff can be nuanced, so always err on the side of discussing it with your favorite trusted security exports - feel free to ask us on the [OpenJS Slack](https://slack-invite.openjsf.org/).

Reviewer: (nit)

```suggestion
When responding to a report, you'll want to seriously consider how the behavior may or may not be a vulnerability. Things to weigh include what documentation says, typical community expectations, what the likely attack surface looks like, etc. These considerations can be nuanced, so always err on the side of discussing it with your favorite trusted security exports - feel free to ask us on the [OpenJS Slack](https://slack-invite.openjsf.org/).
```

> When publishing your fixes, be sure to _credit your reporters_. This can include an explicit reference in the issue and at-mentioning their username(s) in the changelog or in Github Releases. Be sure to wait until the issue has been publicly disclosed before referencing a CVE number or another vulnerability identifier.

Reviewer: consider mentioning there's a built-in mechanism for this in GH PVR
